chain.go

```diff
@@ -106,3 +110,94 @@ func (cc *Conn) AddChain(c *Chain) *Chain {
 
 	return c
 }
+
+// ListChains returns currently configured chains in the kernel
+func (cc *Conn) ListChains() ([]*Chain, error) {
+	conn, err := cc.dialNetlink()
+	if err != nil {
+		return nil, err
+	}
+
+	defer conn.Close()
+
+	msg := netlink.Message{
+		Header: netlink.Header{
+			Type:  netlink.HeaderType((unix.NFNL_SUBSYS_NFTABLES << 8) | unix.NFT_MSG_GETCHAIN),
+			Flags: netlink.Request | netlink.Dump,
+		},
+		Data: extraHeader(uint8(unix.AF_UNSPEC), 0),
+	}
+
+	response, err := conn.Execute(msg)
+	if err != nil {
+		return nil, err
+	}
+
+	var chains []*Chain
+	for _, m := range response {
+		c, err := chainFromMsg(m)
+		if err != nil {
+			return nil, err
+		}
+
+		chains = append(chains, c)
+	}
+
+	return chains, nil
+}
+
+func chainFromMsg(msg netlink.Message) (*Chain, error) {
+	if got, want := msg.Header.Type, chainHeaderType; got != want {
+		return nil, fmt.Errorf("unexpected header type: got %v, want %v", got, want)
+	}
+
+	var c Chain
+
+	ad, err := netlink.NewAttributeDecoder(msg.Data[4:])
```

Comment:
What are you skipping with `msg.Data[4:]`? Is this always safe, or could msg.Data be shorter than 4?

This deserves a comment in the code.

Reply:
I couldn't find out, sorry. It's this way in tables.go too.

Reply:
My best guess is that the 4 bytes are NFTL_CHAIN_FAMILY which is not part of nft_chain_attributes so we need to omit them to parse the attributes. No idea why the response would start with a non-attribute though and so couldn't confirm 100%.